File: node/node_ws/src/demo_cpp_pkg/src/learn_thread.cpp
#include <iostream>
#include <thread> // 线程相关
#include <chrono> // 时间相关
#include <functional>
#include "cpp-httplib/httplib.h" // HTTP客户端库

class DownLoad
{
private:
    /* data */
public:
    void download(const std::string &host, const std::string &path,
                  const std::function<void(const std::string &, const std::string &)> &callback)
    {
        std::cout << "线程 " << std::this_thread::get_id() << " 开始下载 " << host + path << std::endl;
        httplib::Client cli(host.c_str());
        auto res = cli.Get(path.c_str());
        
        // 延迟2秒模拟下载时间
        // std::this_thread::sleep_for(std::chrono::seconds(2));

        if (res && res->status == 200) {
            callback(host + path, res->body);
        } else {
            std::cerr << "Failed to download " << host + path << std::endl;
        }
    }

    void download_start(const std::string &host, const std::string &path,
                        const std::function<void(const std::string &, const std::string &)> &callback)
    {
        std::thread download_thread(&DownLoad::download, this, host, path, callback);
        download_thread.detach(); // 分离线程，允许后台运行
    }
};

int main(){
    auto dl = DownLoad();
    auto word_count = [](const std::string &url, const std::string &content) ->void{
        std::cout << "Downloaded from " << url << ", content size: " << content.length() << " ->" << content.substr(0,9)<< std::endl;
    };

    dl.download_start("http://0.0.0.0:8000", "/novel1.txt", word_count);
    dl.download_start("http://0.0.0.0:8000", "/novel2.txt", word_count);
    dl.download_start("http://0.0.0.0:8000", "/novel3.txt", word_count);

    std::this_thread::sleep_for(std::chrono::seconds(10)); // 主线程等待，确保后台下载线程有时间完成
    return 0;
}

File: node/node_ws/src/demo_python_pkg/demo_python_pkg/learn_thread.py
import threading
import requests

class DownLoad:
    def download(self,url,callback_word_count):
        print(f"线程{threading.get_ident()}开始下载{url}内容")
        response = requests.get(url)
        response.encoding = 'utf-8'
        result = response.text
        callback_word_count(url,result)

    def download_start(self,url,callback_word_count):
        thread = threading.Thread(target = self.download,args=(url,callback_word_count))
        thread.start()


def word_count(url,result):
    print(f"{url}的字数是{len(result)}")


def main():
    download = DownLoad()
    download.download_start("http://0.0.0.0:8000/novel1.txt",word_count)
    download.download_start("http://0.0.0.0:8000/novel2.txt",word_count)
    download.download_start("http://0.0.0.0:8000/novel3.txt",word_count)